abstraction/src/semigroup.rs

```diff
@@ -0,0 +1,18 @@
+use std::ops::Mul;
+
+/// A Semigroup is a set with an associative binary operator.
+/// See [Semigroup on Wikipedia](https://en.wikipedia.org/wiki/Semigroup)
+/// This implementation is inspired by [Haskell's Semigroup typeclass](https://hackage.haskell.org/package/base-4.21.0.0/docs/Data-Semigroup.html)
```

Comment:
Oh, the old `Monoid` docs mentioned that associativity isn't easy to express as a type, so it's up to the implementation